---
title: "MA5111 Assignment"
author: "Oisin Lenaghan"
date: "2023-11-07"
output:
  html_document:
    df_print: paged
---

```{r}
library(dplyr)
library(biomaRt)
library(tximport)
library(rhdf5)
library(gplots)
library(org.Hs.eg.db)
library(DESeq2)
library(DT)
library(apeglm)
library(RColorBrewer)
library(IHW)
library(PCAtools)
library(pheatmap)
library(clusterProfiler)
library(EnhancedVolcano)
library(ComplexHeatmap)
library(circlize)
library(fgsea)
library(tidyverse)
library(ggpubr)
```

# Introduction

This R markdown document walks through an RNA-Seq differential expression analysis using DESeq2 for MA5111. Quality control plots, differential expression results, plots and pathway analysis are covered in the document.
```{r}
quant_dir <- "/home/rstudio/"
list.files(quant_dir)
```
In this R chunk, a directory path is assigned to the variable quant_dir, and then the list.files() function is used to display the names of files in that directory.

## Reading Input Data
```{r}
samples <- read.csv(paste0(quant_dir, "/samples.csv"), header=T, row.names = "Sample", stringsAsFactors = T)
samples
```
This R chunk reads a CSV file named "samples.csv" from the directory specified by quant_dir. The read.csv() function is used to read the CSV file, and the resulting data frame is assigned to the variable samples. The contents of the samples data frame are then displayed.

```{r}
samples$Replicate <- factor(samples$Replicate)

factor_cols <- sapply(samples, is.factor)
factor_cols
```
In this R chunk, the 'Replicate' column in the 'samples' data frame is converted to a factor using the 'factor()' function. The 'sapply()' function is then used to create a logical vector ('factor_cols'), where each element indicates whether the corresponding column in the 'samples' data frame is a factor or not.

```{r}
files <- file.path(quant_dir, rownames(samples), "abundance.tsv")
names(files) <- paste0(rownames(samples))
files
```
This R chunk constructs file paths for each row in the 'samples' data frame using the file.path() function. The resulting vector of file paths is stored in the variable files. Subsequently, names are assigned to the elements in the 'files' vector using row names from the 'samples' data frame. Finally, the resulting vector of file paths with corresponding names is displayed.

```{r}
mart <- useMart(biomart = "ensembl", dataset = "hsapiens_gene_ensembl")
```
In this R chunk, the useMart() function is used to create a connection to the Ensembl Biomart. The Biomart for the Homo sapiens dataset is specified by setting the parameters biomart to "ensembl" and dataset to "hsapiens_gene_ensembl". The resulting connection is stored in the variable mart. This connection can be used to retrieve information about genes and their annotations from the Ensembl database.

```{R, message=F, warning=F}
tx2gene <- getBM(attributes = c("ensembl_transcript_id_version", "hgnc_symbol"), mart = mart, useCache = FALSE)
head(tx2gene)
```
In this R chunk, the getBM() function is used to retrieve information from the Ensembl Biomart. It specifically retrieves Ensembl transcript IDs and their corresponding HGNC symbols. The attributes to be retrieved are specified using the attributes parameter. Messages and warnings are suppressed during the execution of this code by setting the message and warning parameters to FALSE. The resulting data frame is stored in the variable tx2gene, and the first few rows are displayed using the head() function.

```{r}
txi <- tximport(files, type = "kallisto", tx2gene = tx2gene)
head(txi$abundance)
```
In this R chunk, the tximport() function is used to combine transcript-level abundance estimates obtained from Kallisto. The files vector contains the paths to the abundance files generated by Kallisto. The type parameter specifies the quantification method, which is set to "kallisto" in this case. The tx2gene parameter provides the mapping between transcript IDs and gene symbols, which is essential for summarizing the transcript-level information at the gene level. The resulting combined abundance matrix is stored in the txi object, and the first few rows of the abundance matrix are displayed using head(txi$abundance).

```{r}
dds <- DESeqDataSetFromTximport(txi, colData = samples, design = ~ Replicate + Condition )
```
In this R chunk, a DESeqDataSet object is created from the output of tximport using the DESeqDataSetFromTximport function. The txi object contains the combined transcript-level abundance estimates. The colData parameter specifies the sample information from the 'samples' data frame, and the design parameter specifies the experimental design for the differential expression analysis. In this case, the design formula includes the factors 'Replicate' and 'Condition'. The resulting DESeqDataSet object is stored in the variable dds.

```{r}
dds$Condition <- relevel(dds$Condition, ref = "control")
dds <- DESeq(dds)
resultsNames(dds)
```
In this R chunk, the reference level of the 'Condition' factor in the dds (DESeqDataSet) object is set to "control" using the relevel function. This step is often done to establish a baseline level for comparison in differential expression analysis. After releveling, the DESeq() function is used to perform differential expression analysis on the dds object. Finally, the resultsNames() function is employed to display the names of the results available in the DESeq object, which typically include information about log2 fold changes, p-values, and other statistics for each gene.

```{r}
dds <- DESeq(dds)
resultsNames(dds)
```
In this R chunk, additional or updated differential expression analysis is performed on the dds object using the DESeq() function. This step may be redundant or used to ensure that the analysis is applied with the desired settings. The resultsNames() function is then used to display the names of the results available in the updated DESeq object. This typically includes information about log2 fold changes, p-values, and other statistics for each gene.

```{r}
counts <- counts(dds, normalized=TRUE)
```
```{r}
log2_counts <- assay(normTransform(dds))
rld_counts <- assay(rlog(dds))

library(vsn)
library(hexbin)

log2_plt <- meanSdPlot(log2_counts, ranks=FALSE, plot=FALSE)
log2_plt$gg + ggtitle("Log2 + PC Transformation") + xlim(0,20)
```
In the first R chunk, normalized counts are extracted from the DESeqDataSet object 'dds' using the counts() function with the parameter normalized set to TRUE.

In the second R code chunk, log2-transformed counts (log2_counts) and regularized log-transformed counts (rld_counts) are extracted from 'dds' using the normTransform() and rlog() functions, respectively. The meanSdPlot() function from the 'hexbin' package is then used to generate a mean-sd plot for the log2-transformed counts. The resulting plot is customized with a title and x-axis limits using the 'ggplot2' functions.

Log2-transformed gene counts are represented by the log2 transformation plot. The data points represent the mean gene expression for each gene, which is displayed on the horizontal X-axis. In the meantime, the standard deviation of the levels of gene expression is shown on the vertical Y-axis.

Genes located in the vicinity of the lower-left quadrant of the X-axis have modest levels of expression and little fluctuation in their mean expression values. Higher expression levels are seen in the genes that are closer to the lower-right section of the X-axis. On the other hand, genes closer to the top of the plot have higher variability, which is indicated by their higher standard deviation values.

The relationship between the dataset's mean number of RNA reads and the standard deviation of gene read counts is plotted. Comparing the more common genes with a higher average count to the more uncommon, or "once-off," genes in the dataset, this visual representation shows that the latter are more prevalent.

```{r}
rld_plt <- meanSdPlot(rld_counts, ranks=FALSE, plot=FALSE)
rld_plt$gg + ggtitle("Rlog Transformation") + xlim(0,20)
```
In this R chunk, a mean-sd plot is generated for the regularized log-transformed counts (rld_counts) using the meanSdPlot() function from the 'hexbin' package. The resulting plot is then customized with a title ("Rlog Transformation") and x-axis limits using the 'ggplot2' functions. This plot is useful for visualizing the relationship between the mean and standard deviation of gene expression values after regularized log transformation.

The Y-axis in the rlog-transformed graphic shows the standard deviation (SD) of each gene's expression, while the X-axis shows the mean expression level for each gene. The rlog plot, in contrast to the log2-transformed data, shows a drop in the number of genes in the lower-left region of the X-axis, indicating that a greater proportion of genes are highly expressed and fewer genes exhibit low expression levels. The genes' mean and standard deviation values have both significantly increased, indicating that the levels of gene expression have been standardised in this analysis and are now appropriate for more research. The mean line's increasing trend suggests that genes with high expression show higher variation and standard deviation. The plot also identifies 8,103 genes. The same information is presented in this presentation, but because the data has been changed to fit a normalised model, accuracy has increased. As a result, it highlights genes that have greater read counts while reducing the impact of rare counts.

```{r}
write.table(counts, "normalised_counts.txt", sep="\t", quote = F)
write.table(log2_counts, "log2_counts.txt", sep="\t", quote = F)
```
In this R code chunk, the write.table() function is used to write the normalized counts and log2-transformed counts to two separate tab-separated text files. The data is written to "normalised_counts.txt" and "log2_counts.txt" with tab as the column separator (sep="\t") and quotes around the entries turned off (quote = FALSE). This can be useful for storing the preprocessed gene expression data for further analysis or sharing with others.

```{r}
sampleDists <- dist(t(rld_counts))

sampleDistMatrix <- as.matrix(sampleDists)

colnames(sampleDistMatrix) <- NULL

ann <- data.frame(Condition = samples$Condition)

col <- c("blue", "red")
names(col) <- c("tumour", "control")
ann_col <- list(Condition = col)

## match annotation rownames to distance mat
rownames(ann) <- rownames(sampleDistMatrix)

pheatmap(mat=sampleDistMatrix,
         clustering_distance_rows=sampleDists,
         clustering_distance_cols=sampleDists,
         annotation_col = ann,
         annotation_colors = ann_col,
         col=hcl.colors(100,"GnBu",rev=T))
```
In this R chunk, a heatmap of sample distances is generated using the pheatmap function. The distances are calculated based on the regularized log-transformed counts (rld_counts). The colors and annotation for conditions ('tumour' and 'control') are specified, and the heatmap is customized using various parameters such as clustering distances, annotation, and color scales. The resulting heatmap provides a visual representation of the similarity/dissimilarity between samples based on their gene expression profiles.

The cells in the heatmap are colour coded. Genes are shown as rows, and samples are shown as columns. The level of gene expression in the relevant sample is reflected in the colour intensity at each junction. On the other hand, genes expressed in healthy controls are inhibited in tumour samples, and genes repressed in healthy controls exhibit increased expression in tumour cells. Some genes express themselves too much in different tumour copies. Interestingly, Tumour3 has the largest percentage of overexpression compared to all other samples, but Tumor2 completely lacks expression for the same collection of genes.

```{r}
p <- pca(rld_counts, metadata = samples)

biplot(p,
       colby = 'Condition',
       colkey = c('tumour'='royalblue', 'control'='red1'),
       hline = 0,
       vline = 0,
       legendPosition = 'right',
       legendLabSize = 12,
       legendIconSize = 8.0,
       title = 'PCA bi-plot',
       subtitle = 'PC1 versus PC2')
```
In this R code chunk, a Principal Component Analysis (PCA) is performed on the regularized log-transformed counts (rld_counts) with sample metadata (samples). The resulting PCA object is used to create a biplot of the first two principal components. The biplot is customized with colors based on the 'Condition' variable, horizontal and vertical lines at zero, and a legend with specified colors and positioning. The title and subtitle are also added to the biplot for better interpretation. The biplot visually represents the relationships between samples in a reduced-dimensional space defined by the top principal components.

The PCA graphic emphasises the significance of the first two principal components, PC1 and PC2, which together account for 44.11% and 28.78% of the variation in the data. With each point denoting a sample, this graphic illustrates sample associations based on gene expression profiles.

"Control" samples 1, 2, and 3 shows that they have low dissimilarity and similar gene expression patterns, forming a tight cluster. As this is going on, samples 1, 2, and 3 of "Tumour" likewise cluster but also exhibit considerable separation, which is indicative of minor variations in their gene expression profiles.

The principal source of variation that differentiates the "Control" and "Tumour" samples is indicated by their separation along the PC1 axis. The genes responsible for this variance might be important in figuring out the biological differences between the two populations.

The increased heterogeneity seen in the tumour samples is consistent with the hypothesis that cancer cells harbour spontaneous mutations. Moreover, the plot's lack of data overlap indicates that this dataset's differential expression analysis is probably going to function really well.

```{r}
tumour_v_ctrl <- results(dds, filterFun=ihw, alpha=0.05, c("Condition", "tumour", "control"))
res1 <- lfcShrink(dds=dds, res=tumour_v_ctrl, coef=4, type="apeglm")
summary(res1)
```
In this R chunk, differential expression analysis results are obtained using the results() function with the independent hypothesis weighting (ihw) method for filtering at a 0.05 significance level. The contrast is specified as "Condition", with "tumour" as the numerator and "control" as the denominator. The log2 fold changes are then shrunk using the lfcShrink() function, specifically applying the apeglm method. The resulting shrunk log2 fold changes are summarized using the summary() function.

The code specifies an alpha of 0.05, which implies that for any given result to be deemed statistically significant for additional analysis, the p-value must fall below this threshold. Because the corrected p-value is less than 0.05, the findings are statistically significant.

```{r}
get_upregulated <- function(df){
    key <- intersect(rownames(df)[which(df$log2FoldChange>=1)], rownames(df)[which(df$pvalue<=0.05)])
  results <- as.data.frame((df)[which(rownames(df) %in% key),])
    return(results)
}
```
This R code defines a function named get_upregulated. The function takes a data frame (df) representing results from a differential expression analysis as input. It identifies genes that meet the specified criteria for upregulation, where the log2 fold change is greater than or equal to 1 and the p-value is less than or equal to 0.05. The function then extracts and returns a data frame containing the results for the identified upregulated genes.

```{r}
get_downregulated <- function(df){
    key <- intersect(rownames(df)[which(df$log2FoldChange<=-1)], rownames(df)[which(df$pvalue<=0.05)])
    results <- as.data.frame((df)[which(rownames(df) %in% key),])
    return(results)
}
```
This R code defines a function named get_downregulated. Similar to the previous function, it takes a data frame (df) representing results from a differential expression analysis as input. The function identifies genes that meet the specified criteria for downregulation, where the log2 fold change is less than or equal to -1, and the p-value is less than or equal to 0.05. The function then extracts and returns a data frame containing the results for the identified downregulated genes.

```{r}
annotate_de_genes <- function(df){

    df$hgnc_symbol <- rownames(df)
    mart <- useMart(biomart = "ensembl", dataset = "hsapiens_gene_ensembl")
    info <- getBM(attributes=c("hgnc_symbol",
                               "ensembl_gene_id_version",
                               "chromosome_name",
                               "start_position",
                               "end_position",
                               "strand",
                               "entrezgene_description"),
                  filters = c("hgnc_symbol"),
                  values = df$hgnc_symbol,
                  mart = mart,
                  useCache=FALSE)

    tmp <- merge(df, info, by="hgnc_symbol")
    tmp$strand <- gsub("-1", "-", tmp$strand)
    tmp$strand <- gsub("1", "+", tmp$strand)
    tmp$hgnc_symbol <- make.names(tmp$hgnc_symbol, unique = T)
    tmp <- tmp[!grepl("CHR", tmp$chromosome_name),]

    output_col <- c("Gene", "Ensembl ID", "Chromosome", "Start", "Stop", "Strand", "Description", "Log2FC", "P-value", "Adj P-value")
    tmp <- subset(tmp, select=c(hgnc_symbol, ensembl_gene_id_version, chromosome_name, start_position, end_position, strand, entrezgene_description, log2FoldChange, pvalue, padj))
    colnames(tmp) <- output_col

    if(min(tmp$Log2FC) > 0){
        tmp <- tmp[order(-tmp$Log2FC),]
    }else{
        tmp <- tmp[order(tmp$Log2FC),]
    }

    return(tmp)

}
```
This R code defines a function named annotate_de_genes. The function takes a data frame (df) representing differentially expressed genes as input. It adds the HGNC symbol as a column, retrieves additional gene information from the Ensembl Biomart, adjusts the strand representation, ensures unique column names, removes entries related to chromosome names, and reorders the data frame based on Log2FC values. The resulting data frame contains annotated information about the differentially expressed genes.

```{r}
de_up <- get_upregulated(as.data.frame(res1))
de_down <- get_downregulated(as.data.frame(res1))
upregulated_genes <- annotate_de_genes(de_up)
downregulated_genes <- annotate_de_genes(de_down)

#confirm these worked
head(upregulated_genes)

write.table(upregulated_genes, "tumour_vs_control_upregulated.txt", sep="\t", row.names=F, quote=F)
write.table(downregulated_genes, "tumour_vs_control_downregulated.txt", sep="\t", row.names=F, quote=F)
```
In this R chunk, the get_upregulated and get_downregulated functions are used to extract upregulated and downregulated genes, respectively, from the shrunk log2 fold change results (res1). The annotate_de_genes function is then applied to add additional information to these sets of genes. Finally, the first few rows of the resulting annotated data frames (upregulated_genes and downregulated_genes) are displayed using head() to confirm the success of the operations.

```{r}
res1 <- na.omit(res1)

min_width <- min(res1$log2FoldChange)
max_width <- max(res1$log2FoldChange)
max_height <- -log10(min(res1[res1$pvalue>0, 5]))

up <- subset(res1, res1$log2FoldChange > 1 & res1$pvalue <= 0.05)
up <- up[order(-up$log2FoldChange),]
up_list <- head(rownames(up), n=10L)

down <- subset(res1, res1$log2FoldChange < -1 & res1$pvalue <= 0.05)
down <- down[order(down$log2FoldChange),]
down_list <- head(rownames(down), n=10L)

plot_top_20 <- c(up_list, down_list)

EnhancedVolcano(res1,
                lab=rownames(res1),
                x="log2FoldChange",
                y="pvalue",
                selectLab=plot_top_20,
                drawConnectors=TRUE,
                legendPosition = "none",
                FCcutoff=1.0,
                pCutoff=0.05,
                title="Volcano Plot",
                subtitle="Tumour vs. Control",
                caption = paste0('Total Genes = ', nrow(res1)),
                xlim=c(min_width, max_width),
                ylim=c(0, max_height))
```
In this R code chunk, the na.omit() function is used to remove rows with missing values in the shrunk log2 fold change results (res1). The script then defines plot parameters based on the log2 fold change values and p-values. It subsets the data to identify the top 10 upregulated and downregulated genes and combines them for plotting. Finally, the EnhancedVolcano() function is used to create a volcano plot with specific annotations and customization based on the shrunk log2 fold change results.

Given the conditions set for this volcano plot, it is clear from the map that there are substantially more upregulated genes than downregulated genes. The dataset's high frequency of elevated genes may point to a particular reaction to the experimental setup. This implies that a predominate increase in gene expression is linked to the experimental conditions.Additionally, the upregulated genes exhibit a larger degree of change, indicating that the tumour groups express them at a higher level than the control groups.

```{r}
res1 <- na.omit(tumour_v_ctrl)

min_width <- min(res1$log2FoldChange)
max_width <- max(res1$log2FoldChange)
max_height <- -log10(min(res1[res1$pvalue>0, 5]))

up <- subset(res1, res1$log2FoldChange > 1 & res1$pvalue <= 0.05)
up <- up[order(-up$log2FoldChange),]
up_list <- head(rownames(up), n=10L)

down <- subset(res1, res1$log2FoldChange < -1 & res1$pvalue <= 0.05)
down <- down[order(down$log2FoldChange),]
down_list <- head(rownames(down), n=10L)

plot_top_20 <- c(up_list, down_list)
EnhancedVolcano(res1,
                      lab=rownames(res1),
                    x="log2FoldChange",
                    y="pvalue",
                    selectLab=plot_top_20,
                    drawConnectors=TRUE,
                    FCcutoff=1.0,
                    pCutoff=0.05,
                    title="Volcano Plot",
                    subtitle="Tumour vs. Control",
                    legendLabSize=8,
                    caption = paste0('Total Genes = ', nrow(res1)))
```
This code performs a differential expression analysis and creates a volcano plot to visualize the results. It identifies the top 10 upregulated and downregulated genes, sets plot parameters, and utilizes the EnhancedVolcano package for customization. The resulting plot highlights significant gene expression changes between the "Tumour" and "Control" conditions.

This volcano plot highlights the significance of the apeglm function. Genes that  exceed a value beyond + or - 10 are considered outliers and should not be considered in the analysis.

```{r}
res1 <- na.omit(res1)
```
This code removes rows with missing values from the differential expression analysis results stored in the variable res1. This step ensures that the subsequent analysis and visualization are performed on complete data.

```{r}
subset <- rld_counts[, 1:6]

up <- rownames(de_up)
down <- rownames(de_down)

key <- c(up, down)
subset <- subset[which(rownames(subset) %in% key),]

mat <- as.matrix(scale(t(subset), center = T))

ann <- data.frame(Condition = c(rep("Control", 3), rep("Tumour", 3)))
rownames(ann) <- rownames(mat)
col <- c("blue", "forestgreen")
names(col) <- c("Control", "Tumour")
ann_col <- list(Condition = col)

pheatmap(t(mat), 
         show_rownames = FALSE,
         annotation_col = ann,
         annotation_colors = ann_col,
         color = hcl.colors(100, "PRGn",rev=F))
```
This R code conducts a subset analysis on the first 6 columns of regularized log-transformed counts (rld_counts). It extracts the names of upregulated and downregulated genes from previously obtained results (de_up and de_down). A key is formed by combining these gene names, and the subset is further refined to include only the rows corresponding to these genes. The data is then transposed, scaled, and converted to a matrix for heatmap plotting. An annotation data frame (ann) is created to label conditions ('Control' and 'Tumour'). The heatmap is generated using the pheatmap function, where row names are hidden, and conditions are annotated with specified colors. The color scale is determined using the 'PRGn' palette in a reversed order. This code facilitates the visualization of expression patterns for selected genes between 'Control' and 'Tumour' conditions.

The associations between the samples are represented by dendrograms on the top and left sides of the heatmap. The way that samples are grouped according to their gene expression profiles is shown by these hierarchical clustering structures. Annotations for samples are incorporated into the heatmap and help with condition identification by classifying samples into "tumour" (shown in red) and "control" (shown in forest green) conditions. Darker colours on the heatmap indicate a larger degree of dissimilarity, while lighter colours indicate a lower degree of dissimilarity. The dissimilarity values, which are standardised within this range, are represented by a vertical colour bar with values ranging from 0 to 20.

The "tumour" samples stand out due to their intensely coloured hue, which suggests a greater degree of dissimilarity than the "control" samples. The dark colours corresponding to the "tumour" samples show different gene expression profiles, indicating a significant level of variation between them. The larger the dissimilarity with respect to other samples, as indicated by a darker colour in this heatmap. Similarity across sample sets is also seen in the two phylogenetic tree architectures. A set is increasingly alike the farther it branches from another. Here, "tumour1" is significantly different from "tumour2" and "tumour3," suggesting a closer relationship to the control group.

"Tumour3" shows the greatest difference from "control1" on the heatmap when compared to the other control samples. Furthermore, "tumour3" exhibits a comparable degree of dissimilarity with "tumour1."

```{r}
res <- as.data.frame(res1) # tumour vs control
res$hgnc_symbol <- rownames(res)

fgsea_rank <- res %>%
              dplyr::select(hgnc_symbol, log2FoldChange) %>%
              na.omit() %>%
              distinct() %>%
              group_by(hgnc_symbol) %>%
              summarize(stat=mean(log2FoldChange))

fgsea_rank
```
This R chunk conducts preprocessing for gene set enrichment analysis (GSEA). It converts the differential expression analysis results (res1) to a data frame (res). The column 'hgnc_symbol' is added to store HGNC symbols for genes. Subsequently, a ranked gene list (fgsea_rank) is generated by selecting and averaging the log2 fold change values for each unique gene, excluding any missing values. This list serves as input for GSEA, providing a summary statistic for each gene to assess its differential expression between 'tumour' and 'control' conditions.

```{r}
rank <- deframe(fgsea_rank)
head(rank, 20)
```
This chunk uses the deframe function to convert the summarized gene set enrichment analysis (GSEA) results (fgsea_rank) into a named numeric vector (rank). This vector represents the ranked gene list with HGNC symbols as names and the mean log2 fold change values as numeric values. The head function is then used to display the first 20 elements of this ranked list. This step allows for a quick examination of the top entries in the ranked gene list, providing insights into the genes with the highest average log2 fold changes between the 'tumour' and 'control' conditions.

```{r}
pathway <- gmtPathways("/home/rstudio/c6.all.v2023.2.Hs.symbols.gmt")
head(pathway, 1)
```
In this R code chunk, the gmtPathways function is employed to load gene set pathway information from a GMT file. The file path specified is "/home/rstudio/c6.all.v2023.2.Hs.symbols.gmt". The resulting object, assigned to the variable pathway, contains a list of gene sets with their corresponding genes. The head(pathway, 1) command is then used to display the first entry in this pathway data, offering a glimpse into the structure and content of the loaded gene sets.

```{r}
fgsea <- fgsea(pathways=pathway, stats=rank, nperm=1000)

fgseaResTidy <- fgsea %>%
  as_tibble() %>%
  arrange(desc(NES))

fgseaResTidy %>%
  dplyr::select(-leadingEdge, -ES, -nMoreExtreme) %>%
  arrange(padj) %>%
  DT::datatable()
```
In this R code chunk, the fgsea function is utilized to perform gene set enrichment analysis (GSEA). The function takes the loaded pathway information (pathway), the ranked gene list (rank), and the number of permutations (nperm) set to 1000. The results are stored in the fgsea object.

The subsequent code manipulates and visualizes the GSEA results. The fgseaResTidy object is created by converting fgsea to a tibble and arranging the results in descending order based on the normalized enrichment score (NES). The results are then further processed to remove unnecessary columns (leadingEdge, ES, nMoreExtreme). Finally, the results are arranged by adjusted p-values (padj) and displayed as an interactive data table using the DT::datatable() function. This code allows for a comprehensive exploration of enriched gene sets based on the GSEA results, providing insights into pathways associated with differential expression between the 'tumour' and 'control' conditions.

```{r}
filtered_pathway <- subset(fgsea, NES > 1.6)
filtered_pathway
```
In this R code chunk, the subset function is employed to filter the results of the gene set enrichment analysis (GSEA) stored in the fgsea object. Specifically, it selects pathways where the normalized enrichment score (NES) is greater than 1.6, and the filtered results are assigned to the filtered_pathway object. This step allows for focusing on pathways that exhibit a substantial degree of enrichment in the context of the provided gene expression data. The filtered_pathway object is then displayed, providing a list of pathways that meet the specified enrichment score threshold.

The level of overrepresentation of a gene set at the top or bottom of a ranked gene list is indicated by the Enrichment Score (ES). As the analysis moves down the ranked list, the running ES is shown in the top area of the graphic. The gene set's ES is indicated by the peak score that is farthest from 0.0. The most notable gene sets usually have a clear peak at the start or finish of the ranking list. Depending on the direction of the ES, the members that provide the largest contributions to the leading edge subset appear either before or after the peak score.
In this case, the positive value (0.765535) indicates an enrichment toward the top of the ranked list. This gene is at the top of the ranked list.

The NES is 1.631419, suggesting a significant enrichment of the gene set.
The enrichment score's statistical significance is indicated by the p-value. A low p-value (0.00525394) suggests that it is improbable that the enrichment observed happened by accident.

```{r}
filt_up <- as.vector(filtered_pathway$pathway)

for (i in filt_up){
    plt <- plotEnrichment(pathway = pathway[[i]],
    gseaParam = 1, ticksSize = 0.5, stats= rank) +
    labs(title=i) + theme(plot.title = element_text(hjust = 0.5, face="bold"))
    print(plt)
}
```
In this R code chunk, a loop is used to iterate through the filtered pathways stored in the filtered_pathway object. For each pathway (i), the plotEnrichment function is applied to generate an enrichment plot. The function takes the specific pathway information from the pathway object, sets gseaParam to 1, adjusts the size of ticks, and uses the ranked gene list (rank) for statistical analysis. Each resulting plot is then given a title corresponding to the pathway name (i) and displayed. This code facilitates the visualization of enrichment plots for pathways that meet the specified filtering criteria, aiding in the interpretation of the gene set enrichment analysis results.

The rank in the fgsea gene list is indicated by the X-axis for each figure, while the gene set's enrichment score is indicated by the Y-axis. The ES is shown by the green line. The subsets of the gene set list where this specific gene set is expressed and contributes to its positive ES are shown by the black bars.

HINATA_NFKB_IMMU_INF: Around rank 2000 on the x-axis, the red dotted line is reached at an enrichment score of 0.77 as the green line rises at a steep angle. From there, it drops until it reaches the red line at about rank 21000 and enrichment score 0. pointing to a noteworthy pattern of enrichment in this pathway. has a NES score, meaning it is highly related in cancer, but doesn't appear to be very significant, with an associated adjusted pvalue of 0.4964974.

```{r}
filtered_pathway <- subset(fgsea, NES < -1.4)
filtered_pathway
```
In this R code chunk, the `subset` function is used to filter the results of gene set enrichment analysis (GSEA) stored in the `fgsea` object. Specifically, it selects pathways where the normalized enrichment score (NES) is less than -1.4, and the filtered results are assigned to the `filtered_pathway` object. This step allows for focusing on pathways that exhibit substantial downregulation in the context of the provided gene expression data. The `filtered_pathway` object is then displayed, providing a list of pathways that meet the specified enrichment score threshold.

ES: The negative value (-0.3957005) indicates an enrichment toward the bottom of the ranked list.
NES: is -1.425185, suggesting a significant enrichment of the gene set, albeit in the opposite direction as the raw ES.
The p-value represents the statistical significance of the enrichment score. A low p-value (0.003154574) indicates that the observed enrichment is unlikely to occur by random chance.

```{r}
filt_down <- as.vector(filtered_pathway$pathway)

for (i in filt_down){
    plt <- plotEnrichment(pathway = pathway[[i]],
    gseaParam = 1, ticksSize = 0.5, stats= rank) +
    labs(title=i) + theme(plot.title = element_text(hjust = 0.5, face="bold"))
    print(plt)
}
```
In this R code chunk, a loop is utilized to iterate through the pathways that were filtered for downregulation, as specified by the `filtered_pathway` object. For each pathway (`i`), the `plotEnrichment` function is applied to generate an enrichment plot. The function uses the specific pathway information from the `pathway` object, sets `gseaParam` to 1, adjusts the size of ticks, and utilizes the ranked gene list (`rank`) for statistical analysis. Each resulting plot is then given a title corresponding to the pathway name (`i`) and displayed. This code facilitates the visualization of enrichment plots for pathways that exhibit downregulation, aiding in the interpretation of the gene set enrichment analysis results.
